test/e2e/metadata_api_test.go

```diff
@@ -66,32 +71,63 @@ func TestMetadataAPI_Fanout(t *testing.T) {
 	testutil.Ok(t, q.WaitSumMetricsWithOptions(e2e.Equals(2), []string{"thanos_store_nodes_grpc_connections"}, e2e.WaitMissingMetrics))
 	testutil.Ok(t, q.WaitSumMetricsWithOptions(e2e.Equals(2), []string{"thanos_query_metadata_apis_dns_provider_results"}, e2e.WaitMissingMetrics))
 
-	promMeta, err := promclient.NewDefaultClient().MetadataInGRPC(ctx, mustURLParse(t, "http://"+prom1.HTTPEndpoint()), "", -1)
-	testutil.Ok(t, err)
+	var (
+		promMeta map[string][]metadatapb.Meta
+	)
+	// Wait metadata response to be ready as Prometheus gets metadata after scrape.
+	testutil.Ok(t, runutil.Retry(3*time.Second, ctx.Done(), func() error {
+		promMeta, err = promclient.NewDefaultClient().MetadataInGRPC(ctx, mustURLParse(t, "http://"+prom1.HTTPEndpoint()), "", -1)
+		testutil.Ok(t, err)
+		if len(promMeta) > 0 {
+			return nil
+		}
+		return fmt.Errorf("empty metadata response from Prometheus")
+	}))
 
 	thanosMeta, err := promclient.NewDefaultClient().MetadataInGRPC(ctx, mustURLParse(t, "http://"+q.HTTPEndpoint()), "", -1)
 	testutil.Ok(t, err)
+	testutil.Assert(t, len(thanosMeta) > 0, "got empty metadata response from Thanos")
 
 	// Metadata response from Prometheus and Thanos Querier should be the same after deduplication.
-	testutil.Equals(t, thanosMeta, promMeta)
+	metadataEqual(t, thanosMeta, promMeta)
 
 	// We only expect to see one metadata returned.
 	thanosMeta, err = promclient.NewDefaultClient().MetadataInGRPC(ctx, mustURLParse(t, "http://"+q.HTTPEndpoint()), "", 1)
 	testutil.Ok(t, err)
-	testutil.Assert(t, true, len(thanosMeta) == 1)
+	testutil.Assert(t, len(thanosMeta) == 1, "expected 1 metadata from Thanos, got %d", len(thanosMeta))
```

Comment:
What's the reasoning behind these? Why not just `testutil.Equals(t, len(thanosMeta), 1)` and that's it? :P

Reply:
Umm, that' so true! :p I will update it.